File: packages/organization-graph/src/identity.ts
import { createHash } from 'node:crypto';
import type { AuthorityRef } from './types.js';

function sha256(value: unknown): string {
  return createHash('sha256').update(JSON.stringify(value), 'utf8').digest('hex');
}

function authorityIdentity(
  value: Pick<AuthorityRef, 'provider' | 'objectType' | 'objectId'>,
): Pick<AuthorityRef, 'provider' | 'objectType' | 'objectId'> {
  return {
    objectId: value.objectId,
    objectType: value.objectType,
    provider: value.provider,
  };
}

export function deriveGraphNodeId(input: {
  scenarioInstanceId: string;
  type: string;
  authorityRef: Pick<AuthorityRef, 'provider' | 'objectType' | 'objectId'>;
}): string {
  return `node:sha256:${sha256({
    authority: authorityIdentity(input.authorityRef),
    scenarioInstanceId: input.scenarioInstanceId,
    type: input.type,
  })}`;
}

export function deriveGraphEdgeId(input: {
  scenarioInstanceId: string;
  type: string;
  from: string;
  to: string;
  authorityRef?: Pick<AuthorityRef, 'provider' | 'objectType' | 'objectId'>;
}): string {
  return `edge:sha256:${sha256({
    ...(input.authorityRef === undefined
      ? {}
      : { authority: authorityIdentity(input.authorityRef) }),
    from: input.from,
    scenarioInstanceId: input.scenarioInstanceId,
    to: input.to,
    type: input.type,
  })}`;
}

File: services/organization-graph/internal/graphcontract/identity.go
package graphcontract

import (
	"crypto/sha256"
	"encoding/hex"

	"github.com/Negentropy-Laby/OpenSlack/services/organization-graph/internal/graphjson"
)

func digestValue(value graphjson.Value) (string, error) {
	encoded, err := graphjson.Encode(value)
	if err != nil {
		return "", err
	}
	hash := sha256.Sum256(encoded)
	return hex.EncodeToString(hash[:]), nil
}

func authorityIdentity(value AuthorityRef) graphjson.Object {
	return graphjson.Object{
		"objectId": value.ObjectID, "objectType": value.ObjectType, "provider": value.Provider,
	}
}

func DeriveNodeID(scenarioInstanceID, nodeType string, authority AuthorityRef) (string, error) {
	value, err := digestValue(graphjson.Object{
		"authority": authorityIdentity(authority), "scenarioInstanceId": scenarioInstanceID,
		"type": nodeType,
	})
	if err != nil {
		return "", err
	}
	return "node:sha256:" + value, nil
}

func DeriveEdgeID(scenarioInstanceID, edgeType, from, to string, authority *AuthorityRef) (string, error) {
	value := graphjson.Object{
		"from": from, "scenarioInstanceId": scenarioInstanceID, "to": to, "type": edgeType,
	}
	if authority != nil {
		value["authority"] = authorityIdentity(*authority)
	}
	hash, err := digestValue(value)
	if err != nil {
		return "", err
	}
	return "edge:sha256:" + hash, nil
}
